Extensions/MUC-Sub/XMPPMUCSub.m

```diff
@@ -0,0 +1,699 @@
+
+@end
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+@implementation XMPPMUCSub
+
+- (id)init
+{
+    return [self initWithDispatchQueue:nil];
+}
+
+- (id)initWithDispatchQueue:(dispatch_queue_t)queue
+{
+    return [super initWithDispatchQueue:queue];
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+- (BOOL)activate:(XMPPStream *)aXmppStream
+{
+    XMPPLogTrace();
+    
+    if ([super activate:aXmppStream]) {
+        XMPPLogVerbose(@"%@: Activated", THIS_FILE);
+        
+        // xmppStream set by call to super.activate. moduleQueue set by super.init.
+        xmppIDTracker = [[XMPPIDTracker alloc] initWithStream:xmppStream dispatchQueue:moduleQueue];
+        return TRUE;
+    }
+    
+    return FALSE;
+}
+
+- (void)deactivate
+{
+    XMPPLogTrace();
+    
+    dispatch_block_t block = ^{ @autoreleasepool {
+        [xmppIDTracker removeAllIDs];
+        xmppIDTracker = nil;
+    }};
+    
+    if (dispatch_get_specific(moduleQueueTag)) {
+        block();
+    }
+    else {
+        dispatch_sync(moduleQueue, block);
+    }
+    
+    [super deactivate];
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+- (NSString *)supportedBy:(XMPPRoom *)room
+{
+    if (nil == room) {
+        return nil;
+    }
+    
+    // <iq from='hag66@shakespeare.example/pda'
+    //       to='coven@muc.shakespeare.example'
+    //     type='get'
+    //       id='ik3vs715'>
+    //   <query xmlns='http://jabber.org/protocol/disco#info'/>
+    // </iq>
+    
+    NSXMLElement *query = [NSXMLElement elementWithName:@"query" xmlns:XMPPDiscoInfoNamespace];
+    
+    NSString *iqId = [XMPPStream generateUUID];
+    
+    XMPPIQ *iq = [XMPPIQ iqWithType:@"get" elementID:iqId];
+    [iq addAttributeWithName:@"from" stringValue:xmppStream.myJID.bare];
+    [iq addAttributeWithName:@"to"   stringValue:room.roomJID.bare];
+    
+    [iq addChild:query];
+    
+    [xmppIDTracker addElement:iq target:self selector:@selector(handleSupportedByIQ:withInfo:) 
+                      timeout:60];
+    [xmppStream sendElement:iq];
+    
+    return iqId;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+- (NSString *)subscribeTo:(XMPPJID *)room nick:(NSString *)nick password:(NSString *)pass
+{
+    return [self subscribe:xmppStream.myJID to:room nick:nick password:pass];
```

Comment:
What does `xmppStream.myJID` yield if the stream is not active/online?
What do our methods do if the stream is not active/online?

--> Do we need to guard against that?

(Applies to all (?) methods.)

Reply:
Good point. None of the code I've looked at did this so I totally forgot about that myself.

Reply:
Oh my, it _is_ an ... interesting framework.